### PR TITLE
[Remote Store] Add support to create index with remote store by default

### PR DESCRIPTION
**Pull request by linuxpi**
### Description
Allows clusters to be configured for all indices to be Remote Store backed by default. The user doesn't need to specify the settings for remote store and can create an index without specifying Remote Store settings. The cluster will automatically use the default settings and create the index with Remote Store enabled.

### Issues Resolved
[[List any issues this PR will resolve]](https://github.com/opensearch-project/OpenSearch/issues/6861)

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by linuxpi** [comment]
> Why does the cluster setting have the semantics of requiring all new indexes to use remote backed storage, as opposed to defining a default that can be overridden on a per-index basis if desired?

@andrross I've been back and forth b/w the both. The only reasoning behind keeping it this way was to allow users to keep indices in the cluster where remote store is enabled more consistent by enabling this setting.

The approach you mentioned is definitely more user friendly than _forcing_ cluster level settings. Let me think about this a bit more and come to a conclusion.

**Comment by andrross** [comment]
> The only reasoning behind keeping it this way was to allow users to keep indices in the cluster where remote store is enabled more consistent by enabling this setting.

What about indexes created via snapshot restore? Would this do a force override of whatever the previous remote store setting was for the snapshotted index? I'm leaning towards the cluster level setting defining the default, but allowing it to be specifically overridden.

File: distribution/src/config/opensearch.yml [hunk 1]
```diff
@@ -87,6 +87,13 @@ ${path.logs}
 #
 #action.destructive_requires_name: true
 #
+# ---------------------------------- Remote Store -----------------------------------
+# Controls whether cluster imposes index creation only with remote store enabled
+# cluster.remote_store.enabled.force: true
+#
+# Repository to use while enforcing remote store for an index
+# cluster.remote_store.repository.default: my-repo-1
```
Comment (sachinpkale) [inline distribution/src/config/opensearch.yml:95]: Don't we need similar settings for translog as well?

**Comment by sachinpkale** [comment]
Please add tests.

File: distribution/src/config/opensearch.yml (hunk 1 above)
Comment (linuxpi) [reply distribution/src/config/opensearch.yml:95]: i was planning to use the same repo for both segment and translog. but we can add a new setting to provide more flexibility.

**Comment by linuxpi** [comment]
> > The only reasoning behind keeping it this way was to allow users to keep indices in the cluster where remote store is enabled more consistent by enabling this setting.
> 
> What about indexes created via snapshot restore? Would this do a force override of whatever the previous remote store setting was for the snapshotted index? I'm leaning towards the cluster level setting defining the default, but allowing it to be specifically overridden.

yes @andrross, i've update the PR. the cluster settings will act as default if nothing is specified but user will have the option to override these settings.

snapshot restore will act similarly. cluster level settings will be used as default if index metadata doesnt specify these settings

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 2]
```diff
@@ -906,6 +922,43 @@ static Settings aggregateIndexSettings(
...
 
+    /**
+     * Updates index settings to enable remote store by default based on cluster level settings
+     * @param settingsBuilder index settings builder to be updated with relevant settings
+     * @param requestSettings settings passed in during index create request
+     * @param clusterSettings cluster level settings
+     */
+    private static void updateRemoteStoreSettings(Settings.Builder settingsBuilder, Settings requestSettings, Settings clusterSettings) {
+        if (FeatureFlags.isEnabled(FeatureFlags.REMOTE_STORE)
+            && FeatureFlags.isEnabled(FeatureFlags.REPLICATION_TYPE)
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:933]: We need to move these settings under `FEATURE_FLAGGED_CLUSTER_SETTINGS` . In that case, we won't need to put these checks here and user won't be able be set it w/o enabling the feature flags first.

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 3]
```diff
@@ -906,6 +922,43 @@ static Settings aggregateIndexSettings(
...
+            }
+            String segmentStoreRepo = requestSettings.get(SETTING_REMOTE_STORE_REPOSITORY);
+            if (segmentStoreRepo == null) {
+                segmentStoreRepo = CLUSTER_REMOTE_STORE_DEFAULT_SEGMENT_REPO_SETTING.get(clusterSettings);
+            }
+            settingsBuilder.put(SETTING_REPLICATION_TYPE, ReplicationType.SEGMENT)
+                .put(SETTING_REMOTE_STORE_ENABLED, true)
+                .put(SETTING_REMOTE_STORE_REPOSITORY, segmentStoreRepo);
+
+            if (!Objects.equals(requestSettings.get(INDEX_REMOTE_TRANSLOG_STORE_ENABLED_SETTING.getKey()), "false")) {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:950]: we also need to check if `translog_repo` is also supplied . The reason being either we pick both `enabled` and `repo` from index settings . If not provided , fall back to cluster level settings.

**Review by gbbafna — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java (hunk 2 above)
Comment (linuxpi) [reply server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:933]: thanks for pointing this out. i'll update

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 4]
```diff
@@ -906,6 +922,37 @@ static Settings aggregateIndexSettings(
...
+            && FeatureFlags.isEnabled(FeatureFlags.REPLICATION_TYPE)
+            && CLUSTER_REMOTE_STORE_INDEX_FORCE_SETTING.get(clusterSettings)) {
+            // Verify if we can create a remote store based index based on user provided settings
+            if ((INDEX_REPLICATION_TYPE_SETTING.exists(requestSettings)
+                && !INDEX_REPLICATION_TYPE_SETTING.get(requestSettings).equals(ReplicationType.SEGMENT))
+                || (INDEX_REMOTE_STORE_ENABLED_SETTING.exists(requestSettings)
+                    && !INDEX_REMOTE_STORE_ENABLED_SETTING.get(requestSettings))) {
+                return;
+            }
+            String segmentStoreRepo = requestSettings.get(SETTING_REMOTE_STORE_REPOSITORY);
```
[the remark this answers is not in the document]
Comment (linuxpi) [reply server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:942]: updated

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java (hunk 3 above)
Comment (linuxpi) [reply server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:950]: do we not allow user to just send `INDEX_REMOTE_TRANSLOG_STORE_ENABLED_SETTING` as true and pick translog repo from cluster settings? I think we should allow this, let me know what you think

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java (hunk 2 above)
Comment (linuxpi) [reply server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java]: in our case its two features we are dependent on. FEATURE_FLAGGED_CLUSTER_SETTINGS only supports adding settings which are dependent on single feature

File: distribution/src/config/opensearch.yml [hunk 5]
```diff
@@ -87,6 +87,16 @@ ${path.logs}
 #
 #action.destructive_requires_name: true
 #
+# ---------------------------------- Remote Store -----------------------------------
+# Controls whether cluster imposes index creation only with remote store enabled
+# cluster.remote_store.enabled.force: true
```
[the remark this answers is not in the document]
Comment (linuxpi) [reply distribution/src/config/opensearch.yml]: Sure, updated

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 6]
```diff
@@ -906,6 +924,52 @@ static Settings aggregateIndexSettings(
...
+    /**
+     * Updates index settings to enable remote store by default based on cluster level settings
+     * @param settingsBuilder index settings builder to be updated with relevant settings
+     * @param requestSettings settings passed in during index create request
+     * @param clusterSettings cluster level settings
+     */
+    private static void updateRemoteStoreSettings(Settings.Builder settingsBuilder, Settings requestSettings, Settings clusterSettings) {
+        if (CLUSTER_REMOTE_STORE_ENABLED_SETTING.get(clusterSettings)) {
+            // Verify if we can create a remote store based index based on user provided settings
+            if (!canCreateRemoteStoreIndex(requestSettings)) {
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:936]: nit: `canCreateRemoteStoreIndex(requestSettings) == false` (Applicable to all the places where we are using `!` in if condition)

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 7]
```diff
@@ -906,6 +924,52 @@ static Settings aggregateIndexSettings(
...
+                return;
+            }
+
+            settingsBuilder.put(SETTING_REPLICATION_TYPE, ReplicationType.SEGMENT).put(SETTING_REMOTE_STORE_ENABLED, true);
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:940]: Why are we adding Segment Replication setting explicitly here? Are we sure that replication type will always be segment at this point? 
As I understand, Segment replication also has a cluster level setting. `settingsBuilder` should be updated by cluster and index level setting of segment replication prior to `updateRemoteStoreSettings`

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java (hunk 6 above)
Comment (linuxpi) [reply server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java]: Sure

**Pull request merged**